packages/@aws-cdk/aws-codedeploy/lib/deployment-group.ts

```diff
@@ -88,6 +92,10 @@ class ImportedServerDeploymentGroupRef extends ServerDeploymentGroupRef {
         this.deploymentGroupArn = deploymentGroupName2Arn(props.application.applicationName,
             props.deploymentGroupName);
     }
+
+    public get autoScalingGroups(): autoscaling.AutoScalingGroup[] {
```

Comment:
It has to be (the argument to the CodePipeline Deploy Action must be a `ServerDeploymentGroupRef`, exactly like it's `ProjectRef` for CodeBuild, and `RepositoryRef` for CodeCommit). So, for a reference, we simply won't have any ASGs (like we often have an undefined `Role` for reference objects).

Reply:
I see... Why does the deploy action needs access to the ASGs? What would be the degraded experience if I am importing a deployment group?

Reply:
Because it needs to grant each ASG permissions to read from the CodePipeline S3 Bucket. If you don't do this, deployments in that Pipeline will always fail.

Otherwise, the customer has to do this permissions granting themselves, which the L2s are supposed to abstract away.

Reply:
Got it. Thanks!
Maybe worth documenting the motivation for exposing this here.

In such cases we plan to issue some form of notification during synthesis that tells the user that they should make sure some permissions are required. I wonder if you should make this an optional field, so the call site (in this case, the Action), will be able to determine how to deal with the fact that it can't access the list of ASGs. It will be a more accurate representation of that fact than an empty array.

Reply:
I don't hate that idea. The one thing that I don't like is that JSII forces overridden properties to have the same type. So, you can't do `autoScalingGroups?: AutoScalingGroup[]` in `ServerDeploymentGroupRef`, and then `autoScalingGroups: AutoScalingGroup[]` in `ServerDeploymentGroup`. Which means everybody interacting with `ServerDeploymentGroup` will deal with an optional type, while in reality it can never be `undefined`. I'm not sure if that's not too inconveniencing for a small gain.

Let me know what you think.

Reply:
Yes, I realize. But I still think that if this is part of the low level interface, it’s better represented as an optional.

Reply:
OK then. I'll change it to `autoScalingGroups?: AutoScalingGroup[]` in `ServerDeploymentGroupRef`. I'll probably also add `role?: Role` while I'm at it.